File: Go/ExampleTests/common/validvalues_string.go
package common

type ValidValuesString struct {
	Value string
	IsValid string
	Notes string
}

func NewValidValuesStringFromSlice(v []string) ValidValuesString {
	s := ValidValuesString{}
	if len(v) > 0 { s.Value = v[0] }
	if len(v) > 1 { s.IsValid = v[1] }
	if len(v) > 2 { s.Notes = v[2] }
	return s
}

// NewValidValuesStringFromText builds from the text form, e.g. Money as "25 USD".
func NewValidValuesStringFromText(text string) ValidValuesString {
	parts := RequireTokens(text, 3, "ValidValues")
	return ValidValuesString{
		Value: parts[0],
		IsValid: parts[1],
		Notes: parts[2],
	}
}

func (s ValidValuesString) String() string {
	return Token(s.Value) + " " + Token(s.IsValid) + " " + Token(s.Notes)
}

func (s ValidValuesString) Equals(o ValidValuesString) bool {
	return DNCEqual(s.Value, o.Value) &&
		DNCEqual(s.IsValid, o.IsValid) &&
		DNCEqual(s.Notes, o.Notes)
}

func EqualValidValuesStringSlices(a, b []ValidValuesString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/validvalues_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct ValidValuesString {
    pub value: String,
    pub isvalid: String,
    pub notes: String,
}

impl ValidValuesString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            value: v.get(0).copied().unwrap_or("").to_string(),
            isvalid: v.get(1).copied().unwrap_or("").to_string(),
            notes: v.get(2).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 3, "ValidValues");
        Self {
            value: parts[0].clone(),
            isvalid: parts[1].clone(),
            notes: parts[2].clone(),
        }
    }
}

impl std::fmt::Display for ValidValuesString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.value), crate::common::tokens::token(&self.isvalid), crate::common::tokens::token(&self.notes)].join(" "))
    }
}

impl PartialEq for ValidValuesString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.value, &other.value)
            && crate::common::dnc_equal(&self.isvalid, &other.isvalid)
            && crate::common::dnc_equal(&self.notes, &other.notes)
    }
}
